[CI] Failure in SumIT.testSingleValuedFieldWithFormatter

In https://gradle-enterprise.elastic.co/s/zyfrpx2dhy4tq/console-log?anchor=4774  `7.x`

```
REPRODUCE WITH: ./gradlew ':server:integTest' --tests "org.elasticsearch.search.aggregations.metrics.SumIT.testSingleValuedFieldWithFormatter" -Dtests.seed=51FCA3F5A8A9E7FA -Dtests.security.manager=true -Dtests.locale=is-IS -Dtests.timezone=IST -Dcompiler.java=12 -Druntime.java=8
 
```

Fails with some NPE:

```

org.elasticsearch.search.aggregations.metrics.SumIT > testSingleValuedFieldWithFormatter FAILED |  
-- | --
  | Failed to execute phase [dfs_query], |  
  |   |  
  | Caused by: |  
  | java.lang.NullPointerException

```

I found more errors like this one, it does not reproduce. The stack trace looks like:

```
Caused by: java.lang.NullPointerException
	at org.elasticsearch.action.search.SearchProgressListener.searchShards(SearchProgressListener.java:166)
	at org.elasticsearch.action.search.SearchPhaseController$QueryPhaseResultConsumer.consumeInternal(SearchPhaseController.java:639)
	at org.elasticsearch.action.search.SearchPhaseController$QueryPhaseResultConsumer.consumeResult(SearchPhaseController.java:617)
	at org.elasticsearch.action.search.CountedCollector.onResult(CountedCollector.java:62)
	at org.elasticsearch.action.search.DfsQueryPhase$1.innerOnResponse(DfsQueryPhase.java:89)
	... 24 more
```

This seems be related with PR #49471, @jimczi any ideas?

--- a/server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java
+++ b/server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java
@@ -39,6 +39,7 @@ import org.elasticsearch.search.DocValueFormat;
 import org.elasticsearch.search.SearchHit;
 import org.elasticsearch.search.SearchHits;
 import org.elasticsearch.search.SearchPhaseResult;
+import org.elasticsearch.search.SearchShardTarget;
 import org.elasticsearch.search.aggregations.InternalAggregation;
 import org.elasticsearch.search.aggregations.InternalAggregation.ReduceContext;
 import org.elasticsearch.search.aggregations.InternalAggregations;
@@ -564,6 +565,7 @@ public final class SearchPhaseController {
      * iff the buffer is exhausted.
      */
     static final class QueryPhaseResultConsumer extends ArraySearchPhaseResults<SearchPhaseResult> {
+        private final SearchShardTarget[] processedShards;
         private final InternalAggregations[] aggsBuffer;
         private final TopDocs[] topDocsBuffer;
         private final boolean hasAggs;
@@ -600,6 +602,7 @@ public final class SearchPhaseController {
             }
             this.controller = controller;
             this.progressListener = progressListener;
+            this.processedShards = new SearchShardTarget[expectedResultSize];
             // no need to buffer anything if we have less expected results. in this case we don't consume any results ahead of time.
             this.aggsBuffer = new InternalAggregations[hasAggs ? bufferSize : 0];
             this.topDocsBuffer = new TopDocs[hasTopDocs ? bufferSize : 0];
@@ -636,7 +639,7 @@ public final class SearchPhaseController {
                 numReducePhases++;
                 index = 1;
                 if (hasAggs) {
-                    progressListener.notifyPartialReduce(progressListener.searchShards(results.asList()),
+                    progressListener.notifyPartialReduce(progressListener.searchShards(processedShards),
                         topDocsStats.getTotalHits(), aggsBuffer[0], numReducePhases);
                 }
             }
@@ -650,6 +653,7 @@ public final class SearchPhaseController {
                 setShardIndex(topDocs.topDocs, querySearchResult.getShardIndex());
                 topDocsBuffer[i] = topDocs.topDocs;
             }
+            processedShards[querySearchResult.getShardIndex()] = querySearchResult.getSearchShardTarget();
         }
 
         private synchronized List<InternalAggregations> getRemainingAggs() {

--- a/server/src/main/java/org/elasticsearch/action/search/SearchProgressListener.java
+++ b/server/src/main/java/org/elasticsearch/action/search/SearchProgressListener.java
@@ -25,8 +25,10 @@ import org.apache.logging.log4j.message.ParameterizedMessage;
 import org.apache.lucene.search.TotalHits;
 import org.elasticsearch.cluster.routing.GroupShardsIterator;
 import org.elasticsearch.search.SearchPhaseResult;
+import org.elasticsearch.search.SearchShardTarget;
 import org.elasticsearch.search.aggregations.InternalAggregations;
 
+import java.util.Arrays;
 import java.util.List;
 import java.util.Objects;
 import java.util.stream.Collectors;
@@ -165,6 +167,13 @@ abstract class SearchProgressListener {
         return results.stream()
             .filter(Objects::nonNull)
             .map(SearchPhaseResult::getSearchShardTarget)
+            .map(e -> new SearchShard(e.getClusterAlias(), e.getShardId()))
+            .collect(Collectors.toUnmodifiableList());
+    }
+
+    final List<SearchShard> searchShards(SearchShardTarget[] results) {
+        return Arrays.stream(results)
+            .filter(Objects::nonNull)
             .map(e -> new SearchShard(e.getClusterAlias(), e.getShardId()))
             .collect(Collectors.toUnmodifiableList());
     }

